src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp

```diff
@@ -34,108 +41,329 @@ namespace ErrorCodes
+{
+    std::unique_lock lock(mutex);
+
+    auto & row_group = row_groups[row_group_idx];
+    chassert(row_group.status == RowGroupState::Status::Running);
+
+    while (true)
     {
-        throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA, "Error while reading Parquet data: {}",
-                               batch.status().ToString());
+        if (is_stopped || row_group.num_pending_chunks >= max_pending_chunks_per_row_group)
+        {
+            row_group.status = RowGroupState::Status::Paused;
+            return;
+        }
+
+        decodeOneChunk(row_group_idx, lock);
+
+        if (row_group.status == RowGroupState::Status::Done)
+            return;
     }
-    if (*batch)
+}
+
+void ParquetBlockInputFormat::decodeOneChunk(size_t row_group_idx, std::unique_lock<std::mutex> & lock)
+{
+    auto & row_group = row_groups[row_group_idx];
+    chassert(row_group.status != RowGroupState::Status::Done);
+    chassert(lock.owns_lock());
+    SCOPE_EXIT({ chassert(lock.owns_lock()); });
+
+    lock.unlock();
+
+    auto end_of_row_group = [&] {
+        row_group.arrow_column_to_ch_column.reset();
+        row_group.record_batch_reader.reset();
+        row_group.file_reader.reset();
+
+        lock.lock();
+        row_group.status = RowGroupState::Status::Done;
+
+        // We may be able to schedule more work now, but can't call scheduleMoreWorkIfNeeded() right
+        // here because we're running on the same thread pool, so it'll deadlock if thread limit is
+        // reached. Wake up generate() instead.
+        condvar.notify_all();
+    };
+
+    if (!row_group.record_batch_reader)
     {
-        auto tmp_table = arrow::Table::FromRecordBatches({*batch});
-        /// If defaults_for_omitted_fields is true, calculate the default values from default expression for omitted fields.
-        /// Otherwise fill the missing columns with zero values of its type.
-        BlockMissingValues * block_missing_values_ptr = format_settings.defaults_for_omitted_fields ? &block_missing_values : nullptr;
-        arrow_column_to_ch_column->arrowTableToCHChunk(res, *tmp_table, (*tmp_table)->num_rows(), block_missing_values_ptr);
+        if (skip_row_groups.contains(static_cast<int>(row_group_idx)))
+        {
+            // Pretend that the row group is empty.
+            // (We could avoid scheduling the row group on a thread in the first place. But the
+            // skip_row_groups feature is mostly unused, so it's better to be a little inefficient
+            // than to add a bunch of extra mostly-dead code for this.)
+            end_of_row_group();
+            return;
+        }
+
+        initializeRowGroupReader(row_group_idx);
     }
-    else
+
+
+    auto batch = row_group.record_batch_reader->Next();
+    if (!batch.ok())
+        throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA, "Error while reading Parquet data: {}", batch.status().ToString());
```

Comment:
This code is executed with unlocked `lock` and above we have `SCOPE_EXIT({ chassert(lock.owns_lock()); });` and assert will fail. I think we can call `end_of_row_group()` before throwing an exception.